wasm/sdk.go

```diff
@@ -0,0 +1,800 @@
+		reject := args[1]
+
+		go func() {
+			err := sdk.CreateReadPool()
+
+			if err != nil {
+				// fmt.Println("get_public_encryption_key_failed: " + err.Error())
+				reject.Invoke(js.ValueOf("get_blobber_failed: " + err.Error()))
+				return
+			}
+
+			responseConstructor := js.Global().Get("Response")
+			response := responseConstructor.New(js.ValueOf(err))
+
+			// Resolve the Promise
+			resolve.Invoke(response)
+		}()
+
+		return nil
+	})
+
+	// Create and return the Promise object
+	promiseConstructor := js.Global().Get("Promise")
+	return promiseConstructor.New(handler)
+}
+
+func AllocFilter(this js.Value, p []js.Value) interface{} {
+	poolStats := p[0].String()
+	allocID := p[1].String()
+
+	var alloc sdk.AllocationPoolStats
+	err := json.Unmarshal([]byte(poolStats), &alloc)
+	if err != nil {
+		return map[string]interface{}{
+			"error": fmt.Sprintf("AllocFilter failed. Reason: %s", err),
+		}
+	}
+	allocFilter := (*sdk.AllocationPoolStats).AllocFilter
+
+	allocFilter(&alloc, allocID)
+	return nil
+}
+
+func ZBOXGetReadPoolInfo(this js.Value, p []js.Value) interface{} {
+	clientID := p[0].String()
+	handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		resolve := args[0]
+		reject := args[1]
+
+		go func() {
+			result, err := sdk.GetReadPoolInfo(clientID)
+
+			if err != nil {
+				// fmt.Println("get_public_encryption_key_failed: " + err.Error())
+				reject.Invoke(js.ValueOf("get_blobber_failed: " + err.Error()))
+				return
+			}
+
+			responseConstructor := js.Global().Get("Response")
+			response := responseConstructor.New(js.ValueOf(result))
+
+			// Resolve the Promise
+			resolve.Invoke(response)
+		}()
+
+		return nil
+	})
+
+	// Create and return the Promise object
+	promiseConstructor := js.Global().Get("Promise")
+	return promiseConstructor.New(handler)
+}
+
+// // ReadPoolLock locks given number of tokes for given duration in read pool.
+func ReadPoolLock(this js.Value, p []js.Value) interface{} {
+	dur, _ := time.ParseDuration(p[0].String()) // time.Duration,
+	allocID := p[1].String()
+	blobberID := p[2].String()
+	tokens, _ := strconv.ParseInt(p[3].String(), 10, 64)
+	fee, _ := strconv.ParseInt(p[4].String(), 10, 64)
+
+	err := sdk.ReadPoolLock(dur, allocID, blobberID, tokens, fee)
+	if err != nil {
+		fmt.Println("Cannot set wallet info")
+	}
+	return err
+}
+
+// // ReadPoolUnlock unlocks tokens in expired read pool
+func ReadPoolUnlock(this js.Value, p []js.Value) interface{} {
+	poolID := p[0].String()
+	fee, _ := strconv.ParseInt(p[1].String(), 10, 64)
+
+	err := sdk.ReadPoolUnlock(poolID, fee)
+	if err != nil {
+		fmt.Println("Cannot set wallet info")
+	}
+	return err
+}
+
+// //
+// // stake pool
+// //
+
+// // GetStakePoolInfo for given client, or, if the given clientID is empty,
+// // for current client of the sdk.
+func ZBOXGetStakePoolInfo(this js.Value, p []js.Value) interface{} {
+	blobberID := p[0].String()
+	handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
```

Comment:
I can see lots of duplicated code especially with `js.FuncOf` . Pls find a way to avoid it

Reply:
Ok noted, I will investigate it further.